### ISSUE TITLE
Gzip Compression & Browser Caching

### ISSUE THREAD

**Issue by mmstick** [SergioBenitez/Rocket#195]
Is there a way to enable gzip compression & browser caching with Rocket? Chrome/Firefox regularly warn me that my site could save about 70KB combined by enabling compression, and that my resources are missing a cache expiration.

**Comment by belst** [comment]
Since the recommended way to deploy rocket is behind a proxy such as ngnix anyway I guess it could manage caching and compression.

**Comment by mmstick** [comment]
@belst I have zero need for a proxy because I already implement page caching support in my web server with Rocket. No need for a middle man.

**Comment by SergioBenitez** [comment]
There is not built-in implementation of this, but you can easily implement it yourself via a custom [Responder](https://api.rocket.rs/rocket/response/trait.Responder.html), likely using [flate2](https://crates.io/crates/flate2). I'd like to have something like this in `contrib` eventually, but unfortunately there's no Rust-only gzip implementation, making integration difficult.

**Comment by mmstick** [comment]
There's the [zopfli crate](https://github.com/carols10cents/zopfli/) which has a 100% Rust gzip compression implementation.

**Comment by SergioBenitez** [comment]
Neat! I didn't know of it. Unfortunately, according to the `zopfli` documentation, it performs:

> very good, but slow, deflate or zlib compression

speed is _very_ important in an interactive setting such as the web.

**Comment by mmstick** [comment]
It's OK if it's not the fastest, given that I can utilize it with my caching state which would only need to perform compression for a specific page/media once every now and then.

**Comment by mmstick** [comment]
Okay, so I managed to quickly put together a solution with Zopfli and it's working perfectly. All of my JS, CSS, and HTML content is being compressed, and that compressed content is stored within my custom content cache which is using the new `State` mechanism.

For HTML pages which I am generating from templates, it looks something like this:

```rust
impl From<Template> for RequestedContent {
    fn from(template: Template) -> RequestedContent {
        let mut output = Vec::with_capacity(8*1024);
        zopfli::compress(&Options::default(), &Format::Gzip, template.to_string().as_bytes(), &mut output).unwrap();
        RequestedContent {
            data: output,
            content_type: ContentType::HTML,
        }
    }
}
```

Whereas my `RequestedContent` struct is as follows:

```rust
#[derive(Debug,Clone)]
pub struct RequestedContent {
    pub data: Vec<u8>,
    pub content_type: ContentType,
}

impl<'a> Responder<'a> for RequestedContent {
    fn respond(self) -> response::Result<'a> {
        Response::build()
            .sized_body(Cursor::new(self.data))
            .raw_header("Content-Encoding", "gzip")
            .header(self.content_type)
            .ok()
    }
}
```

**Comment by echochamber** [comment]
@mmstick Not sure if you intentionally omitted checking the `Accept-Encoding` request header in your rough implementation for simplicity of example. Just wanted point out you want to include a check for it in your final implementation :)

**Comment by SergioBenitez** [comment]
Looks like there is an answer to this question. I'm happy to revisit this once there's a dependency that's easy to depend on for gzip compression. :)

**Comment by Razican** [comment]
This no longer works, as the `Template.to_string()` method no longer exists. Any other option? I would have used the `show()` method, but seems too slow as per the documentation.

**Comment by Razican** [comment]
Any news on this side, @SergioBenitez?

**Comment by SergioBenitez** [comment]
@Razican You can access the current `Request` from a `Responder` implementation, allowing you to call `Template::respond_to()`.

**Comment by tafia** [comment]
I ended up doing this (with the [deflate](https://crates.io/crates/deflate) crate):

```rust
struct CompressedJson<T>(T);

/// Serializes the wrapped value into gzip compressed JSON. Returns a response with Content-Type
/// JSON and a fixed-size body with the serialized value. If serialization
/// fails, an `Err` of `Status::InternalServerError` is returned.
impl<T: Serialize> Responder<'static> for CompressedJson<T> {
    fn respond_to(self, req: &Request) -> response::Result<'static> {

        let json = to_string(&self.0)
            .map_err(|e| {
                error!("JSON failed to serialize: {:?}", e);
                Status::InternalServerError
            })?;

        let headers = req.headers();
        // check if requests accepts gzip encoding
        if headers.contains("Accept") && 
            headers.get("Accept-Encoding").any(|e| e.to_lowercase() == "gzip") {
            let data = ::deflate::deflate_bytes_gzip(json.as_bytes());
            Ok(Response::build()
               .status(Status::Ok)
               .header(ContentType::JSON)
               .raw_header("Content-Encoding", "gzip")
               .sized_body(::std::io::Cursor::new(data))
               .finalize())
        } else {
            Ok(Response::build()
               .status(Status::Ok)
               .header(ContentType::JSON)
               .sized_body(::std::io::Cursor::new(json))
               .finalize())
        }
    }
}
```

I'm happy with the result, thanks again for Rocket.

**Comment by adwhit** [comment]
I just came up against this. Here is my solution, using a Fairing and the `flate2` crate.

```rust
pub struct Gzip;

impl fairing::Fairing for Gzip {
    fn info(&self) -> fairing::Info {
        fairing::Info {
            name: "Gzip compression",
            kind: fairing::Kind::Response,
        }
    }

    fn on_response(&self, request: &Request, response: &mut Response) {
        use flate2::{Compression, FlateReadExt};
        use std::io::{Cursor, Read};
        let headers = request.headers();
        if headers
            .get("Accept-Encoding")
            .any(|e| e.to_lowercase().contains("gzip"))
        {
            response.body_bytes().and_then(|body| {
                let mut enc = body.gz_encode(Compression::Default);
                let mut buf = Vec::with_capacity(body.len());
                enc.read_to_end(&mut buf)
                    .map(|_| {
                        response.set_sized_body(Cursor::new(buf));
                        response.set_raw_header("Content-Encoding", "gzip");
                    })
                    .map_err(|e| eprintln!("{}", e)).ok()
            });
        }
    }
}
```

Test with `curl`:

```
$ curl http://localhost:8000 --silent --write-out "%{size_download}\n" --output /dev/null
627
$ curl http://localhost:8000 --silent -H "Accept-Encoding: gzip, other things" --write-out "%{size_download}\n" --output /dev/null
299
```

**Comment by spawnia** [comment]
After looking at this issue while searching for a way to add compression, i found that the upcoming `0.5` release of rocket_contrib is going to support compression out of the box.

https://github.com/SergioBenitez/Rocket/blob/master/contrib/lib/src/compression/mod.rs

**Comment by Bauxitedev** [comment]
@adwhit This doesn't seem to work anymore with modern versions of flate2 (1.0.20) so I updated it:

```rust

use flate2::{read::GzEncoder, Compression};
use log::*;
use rocket::{fairing, Request, Response};

pub struct Gzip;

impl fairing::Fairing for Gzip {
    fn info(&self) -> fairing::Info {
        fairing::Info {
            name: "Gzip compression",
            kind: fairing::Kind::Response,
        }
    }

    fn on_response(&self, request: &Request, response: &mut Response) {
        use std::io::{Cursor, Read};

        let headers = request.headers();
        if headers
            .get("Accept-Encoding")
            .any(|e| e.to_lowercase().contains("gzip"))
        {
            response.body_bytes().and_then(|body| {
                let mut encoder = GzEncoder::new(&body[..], Compression::fast());

                let mut buf = Vec::with_capacity(body.len());
                encoder.read_to_end(&mut buf).ok()?;

                response.set_sized_body(Cursor::new(buf));
                response.set_raw_header("Content-Encoding", "gzip");

                Some(())
            });
        }
    }
}
```

**Comment by 18601673727** [comment]
@Bauxitedev thanks a lot for the help, but on `0.5.0-dev` the following one works:
```rust
use flate2::{read::GzEncoder, Compression};
use rocket::{fairing, Request, Response};

pub struct Gzip;

#[rocket::async_trait]
impl fairing::Fairing for Gzip {
    fn info(&self) -> fairing::Info {
        fairing::Info {
            name: "Gzip compression",
            kind: fairing::Kind::Response,
        }
    }

    async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
        use std::io::{Cursor, Read};

        if request.headers().get("Accept-Encoding").any(|e| e.to_lowercase().contains("gzip")) {
            let body_bytes = response.body_mut().to_bytes().await.unwrap();
            let mut encoder = GzEncoder::new(&body_bytes[..], Compression::fast());
            let mut buf = Vec::with_capacity(body_bytes.len());
            let size_read = encoder.read_to_end(&mut buf).unwrap();
            response.set_sized_body(size_read, Cursor::new(buf));
            response.set_raw_header("Content-Encoding", "gzip");
        }
    }
}
```

**Comment by SergioBenitez** [comment]
Please note that calling `body_bytes()` as-is in a response fairing is _always_ wrong. You're reading the entire response body into a buffer. This has horrible performance side-effects, surely negating any positive effects that would result from compression in the common case. And, more importantly, you're creating a trivial DoS attack vector. If clients can force you into large responses, they can exhaust your memory resources. Or, you can DoS yourself. If you ever return an infinite stream, for example, you'll immediately crash your application.

The only correct thing to do is to return a steam that compresses on the fly.

**Comment by 18601673727** [comment]
> Please note that calling `body_bytes()` as-is in a response fairing is _always_ wrong. You're reading the entire response body into a buffer. This has horrible performance side-effects, surely negating any positive effects that would result from compression in the common case. And, more importantly, you're creating a trivial DoS attack vector. If clients can force you into large responses, they can exhaust your memory resources. Or, you can DoS yourself. If you ever return an infinite stream, for example, you'll immediately crash your application.
> 
> The only correct thing to do is to return a steam that compresses on the fly.

Nice framework sir, could you please show how exactly shall us do to enable gzip for rocket?

**Comment by SaadiSave** [comment]
> After looking at this issue while searching for a way to add compression, i found that the upcoming `0.5` release of rocket_contrib is going to support compression out of the box.
> 
> https://github.com/SergioBenitez/Rocket/blob/master/contrib/lib/src/compression/mod.rs

That link does not work. Is supporting gzip compression within rocket still on the table?

**Comment by jebrosen** [comment]
> That link does not work. Is supporting gzip compression within rocket still on the table?

Currently, no. It had some problems and was removed: at the very least, it needed to be entirely reworked for async, and for a long time there were not many good async compression libraries available and versions were mismatched all over the place. And there have been other significant changes between rocket 0.4 and 0.5, for instance the configuration system has been redesigned and now uses the `figment` library. [`async-compression`](https://crates.io/crates/async-compression) looks promising today, but I have not seriously explored it yet.

Automatic compression also has impacts on performance, correctness, and in some cases [security](https://en.wikipedia.org/wiki/BREACH), which makes it more difficult to evaluate for inclusion in Rocket.

**Comment by Ameobea** [comment]
> Looks like there is an answer to this question. I'm happy to revisit this once there's a dependency that's easy to depend on for fast compression. :)

I took a shot at implementing the old `Compression` fairing using `async-compression` and it turned out to be shockingly easy.  I think `async-compression` added some additional shims since the last time I tried to do this, and all the necessary traits are already implemented on `Body` for it to be passed directly to the encoders from that library.

My code for that is here: https://github.com/Ameobea/rocket_async_compression/blob/main/src/lib.rs#L142

It's built as an external library, but it is almost entirely based on the code that was originally in `rocket_contrib` so adding it as a helper library in this repository should be trivial as well.

Let me know if that's something you would be interested in, and I can set up the PR for that to kick the process off!
